331-mqtt-to-postgres-ds

larswitte opened the pull request:
Fixes #331: Changes MQTT-to-postgres microservice.

JeremyTheocharis commented on golang/.idea/golang.iml:
```diff
@@ -1,5 +1,6 @@
 <?xml version="1.0" encoding="UTF-8"?>
```
Can you delete this .idea folder altogether?

larswitte replied:
I removed the folder.

JeremyTheocharis commented on golang/cmd/mqtt-to-postgresql/database.go:
```diff
@@ -1909,3 +1999,28 @@ func GetComponentID(assetID int, componentName string) (componentID int) {
 }
+
+// Todo (naming conventions), sachen
+func GetUniqueProductID(aid int, assetID int) (uid int, err error) {
```
This function is perfectly suited for caching. When calling it, always check the cache beforehand (you can find the cache in the internal package `golang/PKG/internal/cache.go`)

larswitte replied:
Good idea, thank you! I implemented a uniqueProductID cache (`GetUniqueProductIDFromCache()`, `StoreUniqueProductIDToCache()`).

larswitte commented:
After this commit b4ea48ba6a346c297f955f861e3e11fa2dcdcc90 `productTag`, `productTagString`, `addParentToChild` and `uniqueProduct` were correctly inserted into the postgresDB (tried out with nodeRed flow simulating two processing steps).
Changes in the db (accessed with DataGrip) were relatively fast (seconds).

**Open points at the commit:**
- Hypertables were disabled
-  the order of the timestamps in the uniqueProductTable were not checked
- the entry `productId` in the uniqueProductTable contained Strings. It should contain a int referencing productID in the productTable

larswitte commented:
Testing this commit 3773ebb12e2d920ac0b2e2f55e6c0ead70839bf4 with the nodeRed simulation did not function. The productTable should need `addProuct` MQTT Messages, one for each product and asset. After sending them, they didn't appear in the `productTable`. It took more then ten minutes after they appeared in the table. Additionally the asset information (in the `addProduct` Messages was directly used to generate new assets in the asset table (meaning the MQTT messages arrived in time, but were not processed). They also (i think) did not appear in the logs regarding the queue directly. Only after they appeared in the postgresDB I could find them in the logs.

larswitte commented:
> Testing this commit [3773ebb](https://github.com/united-manufacturing-hub/united-manufacturing-hub/commit/3773ebb12e2d920ac0b2e2f55e6c0ead70839bf4) with the nodeRed simulation did not function (in the commit I tried fixing the in my comment above stated open points). The productTable should need `addProduct` MQTT Messages, one for each product and asset. After sending them, they didn't appear in the `productTable`. It always took more then ten minutes after they appeared in the table. Additionally the asset information (in the `addProduct` Messages was directly used to generate new assets in the asset table (meaning the MQTT messages arrived in time, but were not processed). They also (i think) did not appear in the logs regarding the queue directly. Only after they appeared in the postgresDB I could find them in the logs.
> 
> I went through the go code and did not find significant differences to the processing of other MQTT-Topics. I tried sending the `addProduct` messages one by one and as a batch - no difference. Multiple times restarted the cluster - no difference.
> I also switched to the united-manufacturing-hub main branch - no difference (`addProduct` Messages still did not go into the postgresDB).
> 
> I cant explain that behavior. Obviously I am missing something.
> 
> **Possible failure points:**
> 
>     * Message from node red not correctly sent: unlikely because messages direcly arrive, in the end result in correct entries in the `productTable`
> 
>     * Problem with the queue: seems like it but I could not make out one difference in the code which might be the reason
> 
>     * Problem with sql insert/message processing: unlikely because the time delay problem seemingly already exists in the queue.
> 
> 
> Different Observation: At one time one `id` in the `assetTable` was missing (weird because `id` is the serial primary key. Probably unrelated but I wanted to mention it.

I changed all files on the test server,  deactivated the check of the uniqueProduct Timestamps and split up the nodered addProduct and AssemblyStation Mqtt testmessage generation.

Now it works correctly. I think in the future it might be a good idea to add an automatic functionality: if a `uniqueProduct` Message arrives and no productID can be found for the product name and asset, a new entry in the `productTable` should be automatically created. With this functionality no additional `addProduct` message would be required before new uniqueProduct types.

JeremyTheocharis commented on deployment/factorycube-server/templates/mqtttopostgresql-statefulset.yaml:
```diff
@@ -46,16 +46,16 @@ spec:
+          {{- if .Values.mqtttopostgresql.image.tag }}
+          image: {{ .Values.mqtttopostgresql.image.repository }}:{{ .Values.mqtttopostgresql.image.tag }}
           {{- else }}
           image: {{ .Values.mqtttopostgresql.image }}:{{ .Chart.AppVersion }}
```
Adjust this line as well 

```yaml
{{ .Values.mqtttopostgresql.image.repository }}
```

JeremyTheocharis commented on deployment/factorycube-server/templates/timescale-configmap.yaml:
```diff
@@ -178,7 +178,58 @@ data:
-    
+
+    -- #297
+    ALTER TABLE uniqueProductTable ADD COLUMN IF NOT EXISTS uniqueProductAlternativeID TEXT NOT NULL;
```
TODO: check migrations with dataset in old model or introduce breaking changes

JeremyTheocharis commented on deployment/factorycube-server/values.yaml:
```diff
@@ -84,7 +84,10 @@ grafanaproxy:
   enabled: true
-  image: unitedmanufacturinghub/mqtt-to-postgresql
+  image: 
+    repository: larswitte/mqtt-to-postgresql
```
use unitedmanufacturinghub as repository

JeremyTheocharis commented on deployment/factorycube-server/values.yaml:
```diff
@@ -1537,7 +1540,7 @@ grafana:
 ### nodered ###
 nodered:
-  enabled: false 
+  enabled: true
```
leave it disabled by default

JeremyTheocharis commented:
Todo: adjust documentation

larswitte commented:
***!!!Breaking change of postgres: documentation change necessary!!!***
